### PR TITLE
Support for AMD's HIP in BLT

### PR DESCRIPTION
**Pull request by noelchalmers**
This pull request adds support for building and linking using source files using AMD's HIP language. Targets should add 'hip' as a dependency and the ENABLE_HIP=On flag should be passed to cmake.

### CODE REVIEW

**Comment by white238** [comment]
Thank you for the contribution.  Unfortunately we are a bit busy at the moment and it will be a bit before we get to this. Just didn't want you to think we were ignoring you.

**Comment by davidbeckingsale** [comment]
Hey @noelchalmers - is it intentional that this PR removes ROCm?

**Comment by noelchalmers** [comment]
No so much removes as just renames. Our internal version of these ports have all been referencing AMD's HIP API so it seemed natural to use the HIP naming over using ROCm in some select places. We can change this if you all would prefer. It's just a matter of style.

**Comment by davidbeckingsale** [comment]
So is `hcc` used for compiling both rocm and hip?

**Comment by noelchalmers** [comment]
hipcc is used for compiling, which is a fancy wrapper for hcc just as HIP's API is just a fancy wrapper for HCC or CUDA.

**Comment by davidbeckingsale** [comment]
Okay, that makes sense - thanks!

**Comment by noelchalmers** [comment]
Hi everyone, any update on this?

**Comment by rhornung67** [comment]
Hi @noelchalmers Many of us have been on travel at various times over the past several weeks and have been slow to respond as a result. We want to double-check with some folks to make sure we avoid any issues with our copyright statement. I'll try to get this resolved next week when I get back.

File: cmake/BLTMacros.cmake [hunk 1]
```diff
@@ -2,6 +2,9 @@
 # other BLT Project Developers. See the top-level COPYRIGHT file for details
 # 
 # SPDX-License-Identifier: (BSD-3-Clause)
+###############################################################################
+# Copyright (c) 2018,2019 Advanced Micro Devices, Inc.
```
Comment (white238) [inline cmake/BLTMacros.cmake]: Please remove this as it is covered in the above copyright and the base COPYRIGHT file.  All contributors retain copyright for their contributions.

File: cmake/BLTMacros.cmake [hunk 2]
```diff
@@ -586,7 +601,20 @@ macro(blt_add_executable)
         message(FATAL_ERROR "blt_add_executable(NAME ${arg_NAME} ...) given with no sources")
     endif()
 
-    add_executable( ${arg_NAME} ${arg_SOURCES} )
+    list(FIND arg_DEPENDS_ON "hip" check_for_hip)
```
Comment (white238) [inline cmake/BLTMacros.cmake]: Can you pull all this hip logic into a separate macro in BLTPrivateMacros.cmake like blt_setup_cuda_target does? This macro is quite complicated and needs effort to stay readable.

File: cmake/thirdparty/FindHIP/run_hipcc.cmake [hunk 3]
```diff
@@ -0,0 +1,191 @@
+###############################################################################
```
Comment (white238) [inline cmake/thirdparty/FindHIP/run_hipcc.cmake]: I dont know which license this is but please follow the BSD-3-Clause license.  cmake/thirdparty/FindROCm.cmake is a good example of a new file added by a contributor.

File: cmake/BLTMacros.cmake (hunk 1 above)
Comment (noelchalmers) [reply cmake/BLTMacros.cmake]: Hi Chris, we got a go-ahead from our legal department for BLT's new copyright system so I've removed the AMD copyright header from each file as requested.

File: cmake/BLTMacros.cmake (hunk 2 above)
Comment (noelchalmers) [reply cmake/BLTMacros.cmake]: I've tried to re-locate any hip logic into separate blt private macros. Let me know if this change work for you.

File: cmake/thirdparty/FindHIP/run_hipcc.cmake (hunk 3 above)
Comment (noelchalmers) [reply cmake/thirdparty/FindHIP/run_hipcc.cmake]: Done

**Comment by noelchalmers** [comment]
Hey @kennyweiss, I've added hip versions of a host-config and the smoke tests, as well as renamed the existing rocm smoke tests and host-config to hcc. I also made some minor changes in the old rocm host-config to ensure it was working before renaming it.

File: cmake/BLTPrivateMacros.cmake [hunk 4]
```diff
@@ -245,6 +245,121 @@ macro(blt_setup_cuda_target)
     endif()
 endmacro(blt_setup_cuda_target)
 
+##------------------------------------------------------------------------------
+## blt_add_hip_library(NAME         <libname>
+##                     SOURCES      [source1 [source2 ...]]
+##                     HEADERS      [header1 [header2 ...]]
+##                     DEPENDS_ON   [dep1 ...]
+##                     LIBRARY_TYPE <STATIC, SHARED, OBJECT, or blank for executables>
```
Comment (kennyweiss) [inline cmake/BLTPrivateMacros.cmake]: > blank for executables

Is this right? For ``blt_add_library``, omitting the library type reverts to the default/global build type.

Wouldn't one want to call ``blt_add_hip_executable``  for executables?

File: cmake/thirdparty/SetupROCm.cmake [hunk 5]
```diff
@@ -7,7 +7,7 @@
 # ROCM
```
Comment (kennyweiss) [inline cmake/thirdparty/SetupROCm.cmake]: Is this file still needed?

File: tests/smoke/CMakeLists.txt [hunk 6]
```diff
@@ -124,25 +124,48 @@ endif()
 ################
 #  ROCM test
```
Comment (kennyweiss) [inline tests/smoke/CMakeLists.txt]: ROCM test -> HCC test ?

File: cmake/BLTOptions.cmake [hunk 7]
```diff
@@ -43,7 +43,8 @@ option(ENABLE_CLANG_CUDA   "Enable Clang's native CUDA support" OFF)
 mark_as_advanced(ENABLE_CLANG_CUDA)
 set(BLT_CLANG_CUDA_ARCH "sm_30" CACHE STRING "Compute architecture to use when generating CUDA code with Clang")
 mark_as_advanced(BLT_CLANG_CUDA_ARCH)
-option(ENABLE_ROCM         "Enable ROCM support" OFF)
+option(ENABLE_HIP         "Enable HIP support" OFF)
+option(ENABLE_HCC         "Enable HCC support" OFF)
 set(BLT_ROCM_ARCH "gfx900" CACHE STRING "gfx architecture to use when generating ROCm code")
```
[the remark this answers is not in the document]
Comment (noelchalmers) [reply cmake/BLTOptions.cmake]: Let me try and answer your second point. AMD has a c++amp-like programming model for targeting our GPUs called HC, and using the hcc compiler. AMD has also developed the HIP API and hipcc compiler, which serves as a thin layer between an application and CUDA or HC. The hipcc compiler detects (or is told) what architecture to compile for and either inlines CUDA calls or equivalent HC functions at compile time. HIP's syntax is almost identical to CUDA's. While this sounds like just a thin portability layer, it has actually become the preferred development approach for folks using or transitioning to AMD's GPU. This could be due to projects already having existing CUDA code, or just developers being familiar with the CUDA programming model in general.

As for your first question, both the hcc and hipcc compiler can take an architecture argument to specify what device code to generate. I can call BLT_ROCM_ARCH whatever you think makes sense style-wise. The switch in general from using ROCM to HCC in some existing parts of BLT can be discussed too, as it's also just a style choice and can be changed/reverted.

[the remark this answers is not in the document]
Comment (kennyweiss) [reply cmake/BLTOptions.cmake]: Thanks @noelchalmers -- that's really helpful.

Based on your description (where you don't mention ROCM in the first paragraph), perhaps blt should only refer to hip and hcc and avoid referring to rocm.  More terms might simplify things for our users. 

What do others think?

File: cmake/BLTPrivateMacros.cmake [hunk 8]
```diff
@@ -245,6 +245,121 @@ macro(blt_setup_cuda_target)
...
+        set(_non_hip_sources)
+        blt_split_source_list_by_language(SOURCES      ${arg_SOURCES}
+                                          C_LIST       _hip_sources
+                                          Fortran_LIST _non_hip_sources)
+
+        set_source_files_properties( ${_hip_sources}
+                                     PROPERTIES
+                                     HIP_SOURCE_PROPERTY_FORMAT TRUE)
+
+        hip_add_executable( ${arg_NAME} ${arg_SOURCES} )
```
Comment (white238) [inline cmake/BLTPrivateMacros.cmake]: Where is this macro/function? Same goes with hip_add_library...

**Comment by white238** [comment]
Since we can't verify this PR works... Can you elaborate on how these changes are being tested?

File: cmake/BLTPrivateMacros.cmake (hunk 8 above)
Comment (white238) [reply cmake/BLTPrivateMacros.cmake]: i retract this statement.. I found it.

File: cmake/BLTPrivateMacros.cmake (hunk 4 above)
Comment (noelchalmers) [reply cmake/BLTPrivateMacros.cmake]: Good catch. I've removed that from this comment.

File: cmake/thirdparty/SetupROCm.cmake (hunk 5 above)
Comment (noelchalmers) [reply cmake/thirdparty/SetupROCm.cmake]: I believe so, for the ENABLE_HCC option. Are you proposing this file should be renamed as well?

File: tests/smoke/CMakeLists.txt (hunk 6 above)
Comment (noelchalmers) [reply tests/smoke/CMakeLists.txt]: Fixed

**Comment by noelchalmers** [comment]
I've been testing mostly on systems with MI-25s and MI-60s using ROCm 2.1 and now 2.2. Is there any chance you all could try it on one of the AMD systems at LLNL? Crayz or Corona?

**Comment by white238** [comment]
Looks good! Thanks Noel for being patient with us and resolving our comments/concerns.  I will merge this in shortly so you guys can go forward with your work...

File: cmake/BLTOptions.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidbeckingsale) [reply cmake/BLTOptions.cmake]: I'm happy with `BLT_ROCM_ARCH` since ROCM is the underlying stack that HCC and HIP are using.

File: cmake/thirdparty/SetupROCm.cmake (hunk 5 above)
Comment (davidbeckingsale) [reply cmake/thirdparty/SetupROCm.cmake]: I think renaming this would be good.

File: cmake/BLTOptions.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (kennyweiss) [reply cmake/BLTOptions.cmake]: Thanks @noelchalmers and @davidbeckingsale

**Comment by noelchalmers** [comment]
I've renamed that file. Sorry about the late update.

**Comment by white238** [comment]
As soon as it passes the automated tests I will merge it.  At some point we will want some tutorial documentation but I don't want to hold this up any longer and one of us can probably handle it.
